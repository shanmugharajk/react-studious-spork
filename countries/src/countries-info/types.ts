export interface ILanguage {
  code: string;
  name: string;
  native: string;
  rtl: boolean;
}

export interface ICountry {
  name: string;
  code: string;
  emoji: string;
  currency: string;
  languages: ILanguage[];
}

export interface IContinent {
  name: string;
  countries: ICountry[];
}

export interface IContinentQueryResult {
  continent: IContinent;
}

export interface ICountryQueryResult {
  country: ICountry;
}
