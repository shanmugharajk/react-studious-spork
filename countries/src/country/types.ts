export interface ILanguage {
  code: string;
  name: string;
  native: string;
  rtl: boolean;
}

export interface ICountryInfo {
  name: string;
  code: string;
  currency: string;
  languages: ILanguage[];
}
