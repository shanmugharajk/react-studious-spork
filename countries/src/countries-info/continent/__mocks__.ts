import { GET_COUNTRIES_CONTINENT_CODE } from "./queries";

export const continentMockData = {
  name: "Asia",
  countries: [
    { code: "AM", name: "Armenia", emoji: "ðŸ‡¦ðŸ‡²" },
    { code: "BH", name: "Bahrain", emoji: "ðŸ‡§ðŸ‡­" },
    { code: "BN", name: "Brunei", emoji: "ðŸ‡§ðŸ‡³" },
    { code: "BT", name: "Bhutan", emoji: "ðŸ‡§ðŸ‡¹" },
    { code: "LA", name: "Laos", emoji: "ðŸ‡±ðŸ‡¦" },
    { code: "LB", name: "Lebanon", emoji: "ðŸ‡±ðŸ‡§" },
    { code: "MO", name: "Macao", emoji: "ðŸ‡²ðŸ‡´" },
    { code: "NP", name: "Nepal", emoji: "ðŸ‡³ðŸ‡µ" },
    { code: "OM", name: "Oman", emoji: "ðŸ‡´ðŸ‡²" },
    { code: "QA", name: "Qatar", emoji: "ðŸ‡¶ðŸ‡¦" },
    { code: "TR", name: "Turkey", emoji: "ðŸ‡¹ðŸ‡·" },
    { code: "TW", name: "Taiwan", emoji: "ðŸ‡¹ðŸ‡¼" },
    { code: "VN", name: "Vietnam", emoji: "ðŸ‡»ðŸ‡³" },
    { code: "YE", name: "Yemen", emoji: "ðŸ‡¾ðŸ‡ª" },
  ],
};

export const getCountriesContinentCodeMock = [
  {
    request: {
      query: GET_COUNTRIES_CONTINENT_CODE,
      variables: {
        code: "AS",
      },
    },
    result: {
      data: { continent: continentMockData },
    },
  },
];

export const getCountriesContinentCodeNoResultsMock = [
  {
    request: {
      query: GET_COUNTRIES_CONTINENT_CODE,
      variables: {
        code: "P",
      },
    },
    result: {
      data: { continent: null },
    },
  },
];

export const getCountriesContinentCodeErrorMock = [
  {
    request: {
      query: GET_COUNTRIES_CONTINENT_CODE,
      variables: {
        code: "AS",
      },
    },
    error: new Error("An error occurred"),
  },
];
