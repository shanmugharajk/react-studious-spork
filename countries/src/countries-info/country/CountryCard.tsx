import * as React from "react";

import { Alert, Info } from "~/components";
import { isNull } from "~/utils";

import type { ICountry } from "~/countries-info/types";

interface IProps {
  country?: ICountry | null;
  error?: string;
  loading?: boolean;
}

export const CountryCard: React.FunctionComponent<IProps> = ({
  country,
  error,
  loading,
}) => {
  if (error) {
    return <Alert>{error}</Alert>;
  }

  if (loading) {
    return <Info>Loading....</Info>;
  }

  // result from api will set this as null.
  if (isNull(country)) {
    return <Info>No results found</Info>;
  }

  // We clear when user changes the input with undefined to immediately clear the
  // previous details shown to avoid confusion.
  if (!country) {
    return null;
  }

  const { name, code, currency, languages } = country;

  return (
    <div className="mt-5 p-4 space-y-3 text-gray-600 text-sm border border-gray-300 rounded-lg">
      <div>Name: {name}</div>
      <div>Code: {code}</div>
      <div>Currency: {currency}</div>
      <div>Languages: {languages?.map((l) => l.name).join(", ")}</div>
    </div>
  );
};
