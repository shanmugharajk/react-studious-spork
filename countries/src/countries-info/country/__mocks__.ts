import { COUNTRIES_INFO_BY_ID } from "./queries";

export const countryMockData = {
  name: "India",
  code: "IN",
  currency: "INR",
  languages: [{ name: "Hindi" }, { name: "English" }],
};

export const getCountryInfoByIdMock = [
  {
    request: {
      query: COUNTRIES_INFO_BY_ID,
      variables: {
        code: "IN",
      },
    },
    result: {
      data: { country: countryMockData },
    },
  },
];

export const getCountryInfoByIdNoResultsMock = [
  {
    request: {
      query: COUNTRIES_INFO_BY_ID,
      variables: {
        code: "KU",
      },
    },
    result: {
      data: { country: null },
    },
  },
];

export const getCountryInfoByIdErrorMock = [
  {
    request: {
      query: COUNTRIES_INFO_BY_ID,
      variables: {
        code: "JU",
      },
    },
    error: new Error("An error occurred"),
  },
];
