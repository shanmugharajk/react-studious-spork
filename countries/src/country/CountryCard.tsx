import * as React from "react";

export const CountryCard: React.FunctionComponent = () => {
  return (
    <div>
      <span className="text-gray-400">
        Please enter country name to see the details
      </span>
    </div>
  );
};
