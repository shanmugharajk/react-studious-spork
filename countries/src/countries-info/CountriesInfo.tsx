import * as React from "react";

import { Country } from "~/country";
import { Continent } from "~/continent";

export const CountriesInfo: React.FunctionComponent = () => {
  return (
    <div className="mx-10 my-5  rounded-lg shadow-md flex justify-evenly border-t-2 border-indigo-800">
      <Country />
      <Continent />
    </div>
  );
};
