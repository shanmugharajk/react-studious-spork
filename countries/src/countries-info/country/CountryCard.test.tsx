import { render } from "@testing-library/react";

import { ICountry } from "~/countries-info/types";

import { CountryCard } from "./CountryCard";
import { countryMockData } from "./__mocks__";

describe("CountryCard", () => {
  it("should render CountryCard", () => {
    const { container } = render(
      <CountryCard country={countryMockData as ICountry} />,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render 'loading...' message", () => {
    const { container } = render(<CountryCard loading />);

    expect(container).toMatchSnapshot();
  });

  it("should render error message", () => {
    const { container } = render(<CountryCard error="An error occurred" />);

    expect(container).toMatchSnapshot();
  });

  it("should render 'No results found' message", () => {
    const { container } = render(<CountryCard country={null} />);

    expect(container).toMatchSnapshot();
  });
});
