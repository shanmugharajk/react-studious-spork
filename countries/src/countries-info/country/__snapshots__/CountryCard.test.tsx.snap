// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryCard should render 'No results found' message 1`] = `
<div>
  <div
    class="mt-5 p-5 bg-gray-50 border border-gray-300 rounded-lg"
  >
    <span
      class="text-gray-500"
    >
      No results found
    </span>
  </div>
</div>
`;

exports[`CountryCard should render 'loading...' message 1`] = `
<div>
  <div
    class="mt-5 p-5 bg-gray-50 border border-gray-300 rounded-lg"
  >
    <span
      class="text-gray-500"
    >
      Loading....
    </span>
  </div>
</div>
`;

exports[`CountryCard should render CountryCard 1`] = `
<div>
  <div
    class="mt-5 p-4 space-y-3 text-gray-600 text-sm border border-gray-300 rounded-lg"
  >
    <div>
      Name: 
      India
    </div>
    <div>
      Code: 
      IN
    </div>
    <div>
      Currency: 
      INR
    </div>
    <div>
      Languages: 
      Hindi, English
    </div>
  </div>
</div>
`;

exports[`CountryCard should render error message 1`] = `
<div>
  <div
    class="mt-5 p-3 bg-red-50 border border-red-500 rounded-lg"
  >
    <span
      class="text-red-400"
    >
      An error occurred
    </span>
  </div>
</div>
`;
